File: 18_transmitter/half_band_filter_test.py
#  _  ___  _   _____ _     _                 
# / |/ _ \/ | |_   _| |__ (_)_ __   __ _ ___ 
# | | | | | |   | | | '_ \| | '_ \ / _` / __|
# | | |_| | |   | | | | | | | | | | (_| \__ \
# |_|\___/|_|   |_| |_| |_|_|_| |_|\__, |___/
#                                  |___/    
#
# Copyright (c) Jonathan P Dawson 2024
# filename: half_band_filter_test.py
# description:
# License: MIT
#

import subprocess
import numpy as np
from matplotlib import pyplot as plt

subprocess.run(["g++",  "half_band_filter_test.cpp", "half_band_filter.cpp", "-o",  "half_band_filter"], check=True)
process = subprocess.run(["./half_band_filter"], check=True, capture_output=True)
audio_list = []
for line in process.stdout.splitlines():
  audio_list.append(int(line.strip()))

plt.plot(audio_list)
plt.show()

complex_data = np.array(audio_list)

spectrum = 20*np.log10(np.abs(np.fft.fftshift(np.fft.fft(np.hanning(len(complex_data)) * complex_data))))
plt.plot(spectrum)
plt.show()

#plt.plot(audio_list)
#plt.plot(magnitude_list)
#plt.plot(phase_list)
  

File: 18_transmitter/test/half_band_filter_test.cpp
//  _  ___  _   _____ _     _
// / |/ _ \/ | |_   _| |__ (_)_ __   __ _ ___
// | | | | | |   | | | '_ \| | '_ \ / _` / __|
// | | |_| | |   | | | | | | | | | | (_| \__ \
// |_|\___/|_|   |_| |_| |_|_|_| |_|\__, |___/
//                                  |___/
//
// Copyright (c) Jonathan P Dawson 2023
// filename: half_band_filter_test.cpp
// description:
// License: MIT
//

#include <cstdio>
#include <math.h>
#include <stdlib.h>

#include "half_band_filter.h"

int main() {

  half_band_filter filter;

  for (uint16_t idx = 0; idx < 1024; idx++) {
    //int16_t audio = idx == 0 ? 32767 : 0;
    int16_t audio = 32767;
    filter.filter(audio);
    printf("%i\n", audio);
  }

  return 0;
}
